### PR TITLE
[Feature, NOMERGE] RLHF Rollouts

### PR DESCRIPTION
**Pull request by tcbegley**
**Do not merge before #1315**

This PR adds a class that can perform rollouts using the `generate` method of HuggingFace transformer models.

### CODE REVIEW

File: test/test_rlhf.py [hunk 1]
```diff
@@ -164,6 +170,193 @@ def test_get_dataloader(
...
+                in_keys={
+                    "input_ids": "input_ids",
+                    "attention_mask": "attention_mask",
+                    "labels": "labels",
+                },
+                out_keys=["logits"] if inference else ["loss", "logits"],
+            )
+        return model
+
+    def init_reward_model(self, device=None):
```
Comment (apbard) [inline test/test_rlhf.py:231]:
```suggestion
    @staticmethod
    def init_reward_model(device=None):
```

File: test/test_rlhf.py [hunk 2]
```diff
@@ -164,6 +170,193 @@ def test_get_dataloader(
...
+            in_keys=["input_ids", "attention_mask"],
+            out_keys=["rewards", "end_scores"],
+        )
+        return model
+
+    @property
+    def _dummy_batch(self):
+        return PromptData.from_tensordict(
+            TensorDict.load_memmap(f"{HERE}/datasets_mini/tldr_batch")
+        )
```
Comment (apbard) [inline test/test_rlhf.py:246]: why not staticmethod?
```suggestion
    @staticmethod
    def _get_dummy_batch():
        return PromptData.from_tensordict(
            TensorDict.load_memmap(f"{HERE}/datasets_mini/tldr_batch")
        )
```

File: test/test_rlhf.py [hunk 3]
```diff
@@ -164,6 +170,193 @@ def test_get_dataloader(
...
+                == torch.diagonal(scores.log_softmax(-1), 0, -2, -1).squeeze()
+            ).all()
+        else:
+            assert (
+                scores_comp.squeeze() == scores.log_softmax(-1)[..., -1].squeeze()
+            ).all()
+
+    def test_generate(self, max_new_tokens=10):
+        model = self._get_rollout_model(max_new_tokens)
+        batch = self._dummy_batch
```
Comment (apbard) [inline test/test_rlhf.py:328]:
```suggestion
        batch = self._get_dummy_batch()
```

File: test/test_rlhf.py [hunk 4]
```diff
@@ -164,6 +170,193 @@ def test_get_dataloader(
...
+        assert generated.shape == torch.Size(
+            [batch_size, batch.input_ids.shape[1] + max_new_tokens]
+        )
+        assert log_probs.shape == torch.Size([batch_size, max_new_tokens, 1])
+        assert (log_probs <= 0).all().item()
+        assert log_ratio.shape == torch.Size([batch_size, max_new_tokens])
+
+    def test_rollout_from_data(self, max_new_tokens=10):
+        model = self._get_rollout_model(max_new_tokens)
+        batch = self._dummy_batch
```
Comment (apbard) [inline test/test_rlhf.py:341]:
```suggestion
        batch = self._get_dummy_batch()
```

File: torchrl/data/rlhf/utils.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+from torchrl.data.rlhf.prompt import PromptData
+from transformers import GenerationConfig
+
+
+class RolloutFromModel:
+    """A class for performing rollouts with language models.
+
+    Args:
+        model (transformers.Transformer): the model to be used. Should have a
```
Comment (apbard) [inline torchrl/data/rlhf/utils.py:21]: shouldn't we state what defines a "language model"? E.g., 

```suggestion
    """A class for performing rollouts with causal language models, i.e., a model that takes in input tokenized text
        and whose task is to predicting the next word in a sentence having read the n previous words.

    Args:
        model (transformers.Transformer): the model to be used. Should have a
```

File: torchrl/data/rlhf/utils.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+    EOS_TOKEN_ID = 50256
+
+    def __init__(self, model, ref_model, reward_model, max_new_tokens=50):
+        self.model = model
+        self.ref_model = ref_model
+        self.reward_model = reward_model
+        self.max_new_tokens = max_new_tokens
+
+    def kl_step(self):
```
Comment (apbard) [inline torchrl/data/rlhf/utils.py:40]: why we have both kl_step and the rollout functions take kl_coef as input?
isn't choosing either of the two approaches enough?

File: torchrl/data/rlhf/utils.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            generated: The prompt together with generated tokens. This can be obtained
+                by calling the ``generate`` method.
+            log_probs: The log probabilities of the generated tokens. Can be obtained by
+                calling the ``generate`` method.
+            log_ratio: The log ratio of the probabilities of the generated tokens
+                according to the generative model and the reference model. Can be
+                obtained by calling the ``generate`` method.
+            kl_coef: Coefficient with which to multiply the KL term before subtracting
+                from the reward.
+        """
```
Comment (apbard) [inline torchrl/data/rlhf/utils.py:69]: Should we also describe the output structure? (e.g. to define what reward_raw, kl, etc. are?

**Review by apbard — APPROVED** [review]
left few comments. LGTM

File: torchrl/data/rlhf/utils.py (hunk 6 above)
Comment (tcbegley) [reply torchrl/data/rlhf/utils.py:40]: Yes it is, we should move the KL scheduler here, but currently that is not defined on this branch hence the placeholder.

File: torchrl/data/rlhf/utils.py (hunk 7 above)
Comment (tcbegley) [reply torchrl/data/rlhf/utils.py:69]: Yes, good idea, I'll add this

File: torchrl/data/rlhf/utils.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        _, end_scores = self.reward_model(
+            input_ids=rollout_generated[:, -1],
+            attention_mask=rollout_attention_mask[:, -1],
+        )
+        _, end_scores_labels = self.reward_model(
+            input_ids=batch.input_ids,
+            attention_mask=batch.attention_mask,
+        )
+        # the reward is zero except for the timestep where we reached a stopping condition
+        clipped_scores = torch.clip(end_scores - end_scores_labels, -10, 10)
```
[the remark this answers is not in the document]
Comment (tcbegley) [reply torchrl/data/rlhf/utils.py:115]: Sounds good, perhaps in the docstring for the class?

File: torchrl/data/rlhf/utils.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+            tokens = []
+            for i in range(self.max_new_tokens + 1):
+                tokens.append(
+                    torch.where(
+                        arange < rindex + i,
+                        row,
+                        self.EOS_TOKEN_ID,
+                    )
+                )
+            rollout_generated.append(torch.stack(tokens))
```
[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/data/rlhf/utils.py:110]: Example for refactoring
```python
masks = torch.ones(generated.shape[0], max_new_tokens+1, generated.shape[-1], dtype=torch.bool)
for i, ri in enumerate(rindex_total):
    masks[i] = masks[i].tril(ri.item()-1)
generated_copy = generated.unsqueeze(1).expand_as(masks).contiguous()
generated_copy[~masks] = EOS_TOKEN_ID
print(generated_copy)
```

A code to compare against:

```python
from tensordict import TensorDict
import torch

batch = 4
time = 5
generated = torch.arange(batch * time).view((batch, time))
rindex_total = torch.randint(time, (batch,))
print(rindex)
max_new_tokens = 3
EOS_TOKEN_ID = 1234

rollout_generated = []
for rindex, row in zip(rindex_total, generated):
    arange = torch.arange(row.shape[0], device=generated.device)
    tokens = []
    for i in range(max_new_tokens + 1):
        tokens.append(
            torch.where(
                arange < rindex + i,
                row,
                EOS_TOKEN_ID,
            )
        )
    rollout_generated.append(torch.stack(tokens))
rollout_generated = torch.stack(rollout_generated)
```

File: torchrl/data/rlhf/utils.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+        # calculate the reward for the finished sequence
+        _, end_scores = self.reward_model(
+            input_ids=rollout_generated[:, -1],
+            attention_mask=rollout_attention_mask[:, -1],
+        )
+        _, end_scores_labels = self.reward_model(
+            input_ids=batch.input_ids,
+            attention_mask=batch.attention_mask,
+        )
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:141]: do we need 2 calls? Can't this be grouped?

File: torchrl/data/rlhf/utils.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        # the reward is zero except for the timestep where we reached a stopping condition
+        clipped_scores = torch.clip(end_scores - end_scores_labels, -10, 10)
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:143]: that value should be a param of the constructor

File: torchrl/data/rlhf/utils.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        """Makes a step in the KL coefficient schedule."""
+        pass
+
+    @torch.no_grad()
+    def rollout_from_data(self, batch, kl_coef=0.1):
+        generated, log_probs, log_ratio = self.generate(batch)
+        return self.create_rollout_td(batch, generated, log_probs, log_ratio, kl_coef)
+
+    @torch.no_grad()
+    def create_rollout_td(self, batch, generated, log_probs, log_ratio, kl_coef=0.1):
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:56]: this is quite long, maybe split in a bunch of sub-private methods?

File: torchrl/modules/models/rlhf.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
+from pathlib import Path
```
Comment (vmoens) [inline torchrl/modules/models/rlhf.py:1]: add header

**Review by vmoens — COMMENTED** [review]
good progress thanks for this

File: torchrl/data/rlhf/utils.py (hunk 10 above)
Comment (tcbegley) [reply torchrl/data/rlhf/utils.py:141]: I tried grouping but it's non-trivial because they're different shapes. We could pad `batch` perhaps?
